cocos/core/assets/simple-texture.ts

```diff
@@ -198,6 +204,33 @@ export class SimpleTexture extends TextureBase {
         this._mipmapLevel = value < 1 ? 1 : value;
     }
 
+    protected _setMipRange (baseLevel: number, maxLevel: number) {
+        this._baseLevel = baseLevel < 0 ? 0 : baseLevel;
+        this._baseLevel = this._baseLevel < this._mipmapLevel ? this._baseLevel : this._mipmapLevel - 1;
+
+        this._maxLevel = maxLevel < this._baseLevel ? this._baseLevel : maxLevel;
+        this._maxLevel = this._maxLevel < this._mipmapLevel ? this._maxLevel : this._mipmapLevel - 1;
+    }
+
+    /**
+     * Set mipmap level range for this texture.
+     * @param baseLevel The base mipmap level.
+     * @param maxLevel The maximum mipmap level.
+     */
+    public setMipRange (baseLevel: number, maxLevel: number) {
+        if (baseLevel > maxLevel) {
+            warn('baseLevel > maxLevel?');
```

Comment:
Please use warnID with more specific message, like module and API indication.

"Texture: setMipRange failed because base level is larger than max level"

@star-e Also, I don't think we should continue logic if level check failed

Reply:
I didn't know about the warnID system, is there any document about it?
I agree, this check can be made more strict.  @pandamicro

Reply:
The following link shows how to add logs within the engine: https://docs.cocos.com/creator/manual/en/scripting/log.html?h=warnid